tests/contracts/testMarket.nim

```diff
@@ -426,19 +438,214 @@ ethersuite "On-Chain Market":
+    let mockProvider = newMockProvider(
+      numberOfBlocks = 10,
+      earliestBlockNumber = 0,
+      earliestBlockTimestamp = 10,
+      timeIntervalBetweenBlocks = 10
+    )
+
+    let (latestBlockNumber, latestTimestamp) =
+      await mockProvider.blockNumberAndTimestamp(BlockTag.latest)
+    
+    let epochTime = latestTimestamp
+
+    let actual = await mockProvider.blockNumberForEpoch(
+      epochTime.truncate(SecondsSince1970))
+
+    check actual == latestBlockNumber
+  
+  test "blockNumberForEpoch finds closest blockNumber for given epoch time":
+    proc createBlockHistory(n: int, blockTime: int):
+        Future[seq[(UInt256, UInt256)]] {.async.} =
+      var blocks: seq[(UInt256, UInt256)] = @[]
+      for _ in 0..<n:
+        await ethProvider.advanceTime(blockTime.u256)
+        let (blockNumber, blockTimestamp) =
+          await ethProvider.blockNumberAndTimestamp(BlockTag.latest)
+        # collect blocknumbers and timestamps
+        blocks.add((blockNumber, blockTimestamp))
+      blocks
+    
+    proc printBlockNumbersAndTimestamps(blocks: seq[(UInt256, UInt256)]) =
+      for (blockNumber, blockTimestamp) in blocks:
+        debug "Block", blockNumber = blockNumber, timestamp = blockTimestamp
+    
+    type Expectations = tuple
+      epochTime: UInt256
+      expectedBlockNumber: UInt256
+    
+    # We want to test that timestamps at the block boundaries, in the middle,
+    # and towards lower and upper part of the range are correctly mapped to
+    # the closest block number.
+    # For example: assume we have the following two blocks with
+    # the corresponding block numbers and timestamps:
+    # block1: (291, 1728436100)
+    # block2: (292, 1728436110)
+    # To test that binary search correctly finds the closest block number,
+    # we will test the following timestamps:
+    # 1728436100 => 291
+    # 1728436104 => 291
+    # 1728436105 => 292
+    # 1728436106 => 292
+    # 1728436110 => 292
+    proc generateExpectations(
+        blocks: seq[(UInt256, UInt256)]): seq[Expectations] =
+      var expectations: seq[Expectations] = @[]
+      for i in 0..<blocks.len - 1:
+        let (startNumber, startTimestamp) = blocks[i]
+        let (endNumber, endTimestamp) = blocks[i + 1]
+        let middleTimestamp = (startTimestamp + endTimestamp) div 2
+        let lowerExpectation = (middleTimestamp - 1, startNumber)
+        expectations.add((startTimestamp, startNumber))
+        expectations.add(lowerExpectation)
+        if middleTimestamp.truncate(int64) - startTimestamp.truncate(int64) <
+            endTimestamp.truncate(int64) - middleTimestamp.truncate(int64):
+          expectations.add((middleTimestamp, startNumber))
+        else:
+          expectations.add((middleTimestamp, endNumber))
+        let higherExpectation = (middleTimestamp + 1, endNumber)
+        expectations.add(higherExpectation)
+        if i == blocks.len - 2:
+          expectations.add((endTimestamp, endNumber))
```

Comment:
Out of curiosity, why is this an expectation only for i == 8?

Reply:
Sure. So in the test we have `10` blocks. For `10` blocks we have `9` pairs and that's why the loop is going using the range `0..<blocks.len - 1` and then the last index we are processing is `blocks.len - 2`, which is `8`. So when processing this last pair (when `i = 8`) we still need to explicitly add the very last expectation corresponding to the block `i + 1`, which would normally be handled in the next iteration.

Reply:
But actually, I just realized that I moved all those tests to `testProvider.nim` and only now I see that I probably forgot to delete them from `testMarket.nim`... Will do it now immediately...